feat(gswd-articles): add puppeteer idea (#25)
asantos00/website: content/blog/getting-started-with-deno/puppeteer-with-deno.md

Together with great documentation and an engaged community, Puppeteer has been one of the best solutions when it comes to writing applications connecting with headless browsers.

A few months ago, [Luca Casonato](https://twitter.com/lcasdev), one of Deno's core team members, ported Puppeteer into Deno.

https://twitter.com/lcasdev/status/1344279906809741312

This is, by itself, an interesting topic to explore: how much did the code have to change to get it working in Deno. But that's not what we're doing here today.

As Deno gains its place among script tools, being able to use *puppeteer* is another interesting step taken. With `deno-puppeteer`, users  can now benefit from the ease of use of Deno while writing *puppeteer* scripts.

In this blogpost we will build a CLI utility that demonstrates that.

The full code is available [here](https://github.com/asantos00/deno-website-diff).
## Writing a Puppeteer script

The objective of the CLI utility we'll build is to check a website for visual changes in different resolutions.

> This tool will compare the website with the last time it checked, creating an image with the difference. It can work as a QA assurance tool, something that runs after deploying a website to confirm that it is working fine in different resolutions.

Reviewer: ```suggestion
This tool will compare the website with the last time it checked, creating an image with the difference. It can work as a QA assurance tool, something that runs after deploying a website to double-check the changes done and that it is working fine in different screen resolutions.
```

> To achieve this, we'll use a couple community packages and tools. Some are functions from Deno's standard library, others are just Node.js packages

Reviewer: ```suggestion
To achieve this, we'll use a couple community packages and tools. Some are functions from Deno's standard library, others are just existing Node.js packages
```
(to emphasize the fact that it leverages existing Node.js packages already known and loved by the community)

- pngjs - PNG encoder/decoder in JS
- pixelmatch - A JS image comparison library
- Deno file-system APIs (`readFile` and `writeFile`)

> We'll use [jspm](https://jspm.org/) to make sure `pngjs` and `pixelmatch` (both Node.js packages) are **ES6 module compatible**. This will make sure they work on Deno (yes, Deno is fully ES6 compatible!).

Reviewer: ```suggestion
We'll use [jspm](https://jspm.org/) to make sure `pngjs` and `pixelmatch` (both Node.js packages) are **ES6 module compatible**. This will make sure that they work on Deno (yes, Deno is fully ES6 compatible!).
```

```

After having a screenshot, it is just a matter or saving it on the filesystem or comparing it with the one previously stored.

To save to the file system we'd just use `Deno.writeFile`. To compare with the previously saved image we'll use `pixelmatch` and `pngjs`, as shown by the script below.

```ts
import pixelmatch from "https://jspm.dev/pixelmatch";
import { PNG } from "https://jspm.dev/pngjs";

// Cut for brevity

const newImage = await parsePNG(screenshot as Uint8Array);
const oldImage = await parsePNG(await Deno.readFile(screenshotPath));
const diff = new PNG(viewPort);

pixelmatch(
  newImage.data,
  oldImage.data,
  diff.data,
  viewPort.width,
  viewPort.height,
  { threshold: 0.5 },
);

const buffer = PNG.sync.write(diff);
await Deno.writeFile(
  // domain is the website domain, resolution the screen resolution
  `${screenshotBasePath}/diff-${domain}-${resolution}.png`,
  buffer,
  { create: true },
);
```

The full code is available [here](https://github.com/asantos00/deno-website-diff).

Again, the code is quite straightforward. We're using Deno runtime APIs to access the filesystem, and the third-party packages to decode and compare both images.

> If the Deno APIs look strange to you, fear nothing, Deno's documentation is quite complete on that.

Reviewer: ```suggestion
If the Deno APIs look strange to you, fear nothing, Deno's documentation is quite detailed and thorough.
```

Deno supports directly importing modules by URL in the code, and the documentation isn't any different. By using the provided URL (https://doc.deno.land) plus the URL to a package, we get a beautifully designed page with the package types and documentation.

For instance, to check `readFile` documentation we just need to navigate to https://doc.deno.land/builtin/stable#Deno.readFile and this is what we get.

![readfile-documentation](./deno-read-file.png)

*This is possible to be used with a direct GitHub link to any TS file, ex: https://doc.deno.land/https/raw.githubusercontent.com/timonson/djwt/master/mod.ts*

## Executing the code

After having Deno installed on the system, executing the code is the simplest of tasks. We can just use `deno run` plus the path to the file we want to execute (in our case we've called it `mod.ts`).

However, and as we previously mentioned, all Deno programs run in a sandbox thus we need to give your program the specific permissions it needs to execute.

For our specific usecase, and because puppeteer needs quite a lot of permissions, it needs to access the environment, network, file system, and have the ability to run processes.

> We'll also have to use the `--unstable` flag to enable our program to access a few unstable Deno APIs.

Reviewer: Maybe interesting to just briefly explain which unstable APIs and the fact that it is due to notify users that these features are still under active development. Just so that people do not get thrown off by the fact that deno is sold as:
- Core security features (but then skip/allow all stuff)
- Nice and well written language (but then flag it as unstable)

Check file:///Users/alexandre/dev/personal/deno/puppeteer/mod.ts
✅ Successfully installed website_compare
/Users/alexandre/.deno/bin/website_compare
```

*Note how we're running the `install` command with the `--name` flag to set the name of the executable. We're also setting the permissions the script needs to execute, this is the only time we'll need to do it.*

After doing this, we can run the `website_compare` wherever we are in our system.

```bash
$ website_compare https://picsum.photos
Diff stored at /Users/alexandre/dev/personal/deno/puppeteer/screenshots/diff-picsum-mobile.png
```

And this completes our objective for the blogpost!

We've demonstrated how can we use Deno's simplicity to make it even easier to write *puppeteer* scripts.

As you've seen everything works the same as in Node.

The big difference is that we can take advantage of some Deno advantages. Those are things like native TypeScript support, a clean standard-library, no *node_modules* and fine-grained permission control.

## Conclusion

Today we've explored `deno-puppeteer`, a package that makes it possible to write *puppeteer* scripts on Deno.

> Scripting is one of many use cases for Deno, and one of the reasons Deno was created, but definitely not the only one. In this blogpost series' we'll explore different parts of Deno, from rust interoperability to static site generation.

Reviewer: ```suggestion
Scripting is one of many use cases for Deno, and one of the reasons Deno was created, but definitely not the only one. In this blog post series' we'll explore different parts of Deno, from rust interoperability to static site generation.
```

We truly believe that, by building on the shoulders of giants (TypeScript, Node.js and Rust) and adding a few things to the mixture, Deno has all it takes to stand out.

> If you're interested in knowing more about Deno and how to use it to build tools and web applications, make sure you checkout my recently launched book [Getting started with Deno](). In the book, we'll carefully explain all the mentioned Deno features (and many others) while building real-world applications.

Reviewer: Just as a reminder that the link to the book is missing, maybe even have a microwebsite created by you presenting the book, redirecting people to the appropriate amazon store (.com, .es, .de, .it, .fr, .com.br, .co.uk, etc.) enabling you to collect more analytics and metrics.

Author: Yeap, that's an ongoing work in another PR #24 but one that isn't finished yet! Thanks for the suggestion!

> This article (and code) was written by me my friend [Felipe Schmitt](https://twitter.com/schmittfelipe), a declared *puppeteer* fan which is always ready to explore new pieces of technology.

Reviewer: ```suggestion
This article (and code) was written by me and my friend [Felipe Schmitt](https://twitter.com/schmittfelipe), a declared *puppeteer* fan which is always ready to explore new pieces of technology.
```
